[ZIP 303] Sprout Payment Disclosure (#119)
zcash/zips: drafts/bitcartel-payment-disclosure/draft.rst

| Title: Payment Disclosure
| ZIP: 00? (DRAFT)
| Replaces: n/a
| Superseded-by: n/a
| Editor: Simon Liu <simon@z.cash>

==================
Payment Disclosure
==================

Abstract
--------

This ZIP describes a method of proving that a payment was sent to a shielded address.  In the typical case, this means enabling a sender to present a proof that they transferred funds to a recipient's shielded address.  The method described will be compatible with the Zcash Protocol [PROTOCOL] deployed at the launch of the Zcash network.

This document follows the common conventions defined for ZIPs in [ZIP-1].

.. [PROTOCOL] Zcash Protocol Specification: https://github.com/zcash/zips/blob/master/protocol/protocol.pdf
.. [ZIP-1] https://github.com/zcash/zips/blob/master/zips/0001.rst

Copyright Notice
----------------
This ZIP is published under the MIT License.

Motivation
----------
Payment disclosure is useful in a number of situations, for example:

* A sender may need to prove that their payment was sent and received by a recipient.  For example, a customer paid too much for an item and would like to claim a refund from the vendor.

> * A third party needs to verify that a payment between sender and recipient was executed successfully.  For example, a regulator needs to confirm a transfer of funds between two parties.

Reviewer: This is broken when the sender and receiver collude (https://github.com/zcash/zcash/issues/2036#issuecomment-329864311).

The ZIP must be clearer on what are the claimed security properties of payment disclosure.

Author: What would you add here?  Payment disclosure proves a payment occurred between two parties, but it doesn't prove that the payment being disclosed is the actual payment a third party is interested in.

Reviewer: The weaker property you describe seems true, but not trivially -- you will need to invoke the binding property of the commitment scheme to prove it. So don't believe it until you formally state the property and sketch the proof.

Reviewer: With the `DecryptNote` fix, maybe the third party _will_ be convinced of the amount as well (checking its consistency with the commitment). Again, this needs a proof.

Reviewer 2: Publishing this zip should not be blocked on a security proof; that should be in a ZIP similar to [310](https://zips.z.cash/zip-0310)

Bob retrieves his payment disclosure and emails it to Alice.  Alice is travelling and only has access to her smartphone.  She visits a Zcash blockchain explorer and copies the payment disclosure into the search box. The explorer confirms that Bob paid his share.

Charlie attempts to retrieve the payment disclosure but can't find one.  Charlie had used the command line to send his funds to Alice, but never checked the status of the z_sendmany operation.  Checking the debug logs he discovers that the operation failed due to an incorrect parameter.  Charlie tells Alice that he will re-send his payment.

Design Considerations
---------------------
The payment disclosure does not prove that the party presenting the payment disclosure is the sender.

To mitigate against a replay attack, the recipient could enforce a policy where the first person to present a given payment disclosure is the honest sender.

To prevent a man-in-the-middle attack, the recipient could pose an interactive challenge involving some out-of-band secret which only the sender would pass.

A non-interactive solution is also possible:

> 

Reviewer 3: I don't see how these protect against man-in-the-middle.
I wonder if it's at all possible to protect against it in non-interactive setting

> 

Author: In the interactive case?  Added a TODO to add an example.

#. The sender includes data, such as a refund address, in the memo field of a payment to a shielded address.  The memo field is non-malleable as it is part of the data signed by the sender when creating JoinSplitSig [PROTOCOL] §4.1.6 Signatures.  Currently, the public key to verify the signature is stored in the blockchain along with the transaction.
#. When presenting the payment disclosure to a recepient, the sender signs the payment disclosure with the JoinSplitSig private key which only they would know.  The JoinSplitSig key pair is ephemeral and normally discarded but the method described in this ZIP will retain the private portion.  The recipient can then use the public key to verify the signed disclosure.

Using the JoinSplitSig key pair in this way also means that a sender can make payments without having to first provide a refund address or other detail at the time of the original payment.  If at later date the sender and recipient need to confirm the payment, the sender can sign and send any data required to the recipient, who can in turn verify it.

> 

Reviewer 3: detail->details

> 

Author: Fixed

Known Issues
------------

> Both plaintext outputs of a joinsplit are currently encrypted with the same ephemeral secret key. (TODO: Is there an existing ticket which explains the rationale for this?)

Author: https://github.com/zcash/zcash/issues/558

> Both plaintext outputs of a joinsplit are currently encrypted with the same ephemeral secret key. (TODO: Is there an existing ticket which explains the rationale for this?)

Author: encrypted with symmetric keys derived from the same ephemeral secret key

> Both plaintext outputs of a joinsplit are currently encrypted with the same ephemeral secret key. (TODO: Is there an existing ticket which explains the rationale for this?)

Author: Added

> Both plaintext outputs of a joinsplit are currently encrypted with the same ephemeral secret key. (TODO: Is there an existing ticket which explains the rationale for this?)

Author: Changed

This means that a payment disclosure which includes the ephemeral secret key intended to decipher the note ciphertext belonging to a particular joinsplit output, creates a privacy leak by also deciphering the other joinsplit output and revealing it's amount and payment address.

Note the current implementation of the ``z_sendmany`` RPC call:

- When sending from a transparent address to shielded addresses, each shielded address will occupy one output of a joinsplit.  Thus a payment address to a different recipient could be revealed.

- When sending involves chained joinsplits, for each joinsplit, one output is for payment to a recipient's address, while the other output is used as change back to the sender.  Thus the change address will be revealed, which is the sender's own shielded address.

A proposal to prevent information leakage of change addresses is currently under development [KDFT].  TODO: Does this block ZIP approval?

> Alao note that an independent third party cannot know for sure if the sender and recipient are not colluding to hide value transfer.  JoinSplits have two inputs and two outputs.  The recipient may identify their receiving address for one output, but not disclose the fact that the other output also belongs to them.

Reviewer 3: Alao->Also

- index [0..1] of JoinSplit output
- recipient's payment address is a shielded address ``(a_pk, pk_enc)`` [PROTOCOL] §3.1 Payment Addresses and Keys
- symmetric key used to encrypt the note, also referred to as the ephemeral secret key ``esk`` [PROTOCOL] §4.10.1 Generate a new KA (public, private) key pair ``(epk, esk)``.
- JoinSplitSig private key used to sign the JoinSplit transaction ``joinSplitPrivKey`` [PROTOCOL] §4.4 Sending Notes

The payment disclosure data should be persisted to disk or a database so it can be retrieved later.

When persisting, third party applications should expect serialization of payment disclosure to follow Zcash and upstream convention.

::

    ADD_SERIALIZE_METHODS;
    ...
    READWRITE(txid);
    ...
    READWRITE(joinSplitPrivKey);

A new RPC call will be introduced to allow the sender to retrieve the payment disclosure data for a given shielded output index.

``z_getpaymentdisclosure txid joinsplit_index output_index [message]``

- Returns the payment disclosure in hexadecimal format.
- Message is an optional parameter, a UTF-8 string.  We may want to restrict/sanitize this user input, e.g. number of characters, allowed characters.

> The sender want sthe payment disclosure to be non-malleable, to prevent an attacker modifying details like the refund address.  To achieve this, the sender will sign the payment disclosure with the JoinSplitSig private key and append the signature to the end of the payment disclosure data.

Reviewer 3: want sthe -> wants the

> The sender want sthe payment disclosure to be non-malleable, to prevent an attacker modifying details like the refund address.  To achieve this, the sender will sign the payment disclosure with the JoinSplitSig private key and append the signature to the end of the payment disclosure data.

Author: Fixed

A new RPC call will be introduced to allow a third party to verify a payment disclosure.

``z_validatepaymentdisclosure paymentdisclosure``

Validates a payment disclosure and returns JSON output as follows:

::

    {
    valid : true,
    rawhex : "0011...ff",
    version: 0
    txid: "4e...",
    jsindex: 0,
    outputindex: 1,
    paymentaddress: zABC123...,
    value: 10.56100002
    message: "pay me at z007",
    signature: "10ab23..."
    }

> Valid field is true if all the following conditions hold:

Reviewer: No checking of the note commitment, as in  [PROTOCOL] §4.10.2?

Author: Now fixed.

* `-paymentdisclosure=1`
* `-experimentalfeatures=1`

A third party trying to validate payment disclosure must track all transactions and thus enable the option:

* `-txindex=1`

In this implementation we will assume that once the feature is enabled, the node will create and store a payment disclosure for every transaction sent which involves a shielded address.

When the sender calls RPC call ``z_sendmany`` and creates a shielded transaction in ``asyncrpcoperation.cpp``:

#. retain the ephemeral joinSplitPrivKey used to sign the transaction
#. retain the ephemeral secret key used for symmetric encryption of the note plaintext

Record relevant data in a struct (or class) defined as follows:

::

    struct PaymentDisclosureKey {
        uint256 txid            // primitives/transaction.h
        size_t js;              // Index into CTransaction.vjoinsplit
        uint8_t n;              // Index into JSDescription fields of length ZC_NUM_JS_OUTPUTS
    };

    struct PaymentDisclosureInfo {
        uint8_t version;          // 0 = experimental, 1 = first production version, etc.
        uint256 esk;              // zcash/NoteEncryption.cpp

>         uint256 joinSplitPrivKey; // primitives/transaction.h

Author: Instead of uint256, may be more convenient to store full 64 byte output from libsodium crypto_sign_keypair() rather than extracting the private key and later having to reconstruct output via tx.joinSplitPubKey.  The full 64 bytes is required by crypto_sign_detached().

>         uint256 joinSplitPrivKey; // primitives/transaction.h

Author: We can easily reconstruct the 64 byte buffer so we will only store the 32 bytes per the Ed25519 RFC https://tools.ietf.org/html/rfc8032 and avoid being tied to an implementation e.g. Libsodium, SUPERCOP which both store 64 bytes, https://blog.mozilla.org/warner/2011/11/29/ed25519-keys/

            ) == 0))
        {
            throw std::runtime_error("crypto_sign_verify_detached failed");
        }

#. Construct and serialize a PaymentDisclosure object using the data generated so far.

    ::
    
        struct PaymentDisclosure {
            PaymentDisclosurePayload    payload;
            unsigned char[64]           payloadSig;
        }

#. Return to the caller a JSON object containing the hex string of the serialized PaymentDisclosure.  If there were errors generating the payment disclosure, return a standard JSON-RPC error with an appropriate error message.

::

    {
      "hex" : "091234ABCDEF234781248273842738491....."
    }

This raw hex string can be given to the recipient or a third party to use with the new RPC call ``z_validatepaymentdisclosure``:

``z_validatepaymentdisclosure paymentdisclosure``

To validate a payment disclosure, perform the following steps:

1. Deserialize the raw hex string into a ``PaymentDisclosure`` object
2. Retrieve the ``joinSplitPubKey`` for the transaction and verify the payment disclosure signature ``payloadSig``.
3. Retrieve the note ciphertext from the blockchain for ``txid``, ``js``, ``n``.

> 4. Use the ``esk`` to decrypt the ciphertext into plain text

Reviewer 3: Are we assuming the decrypter is the recipient? Or know recipient's public key? As this is needed in addition to esk to generate K_i^enc needed for decryption of note

Reviewer: For clarity, explicitly specify the extra input?

Reviewer 4: It's not an extra input, it's part of the payment disclosure.

> 4. Use the ``esk`` to decrypt the ciphertext into plain text

Author: From most recent discussion, the payment disclosure itself will include the recipient's public key.

> 4. Use the ``esk`` to decrypt the ciphertext into plain text

Reviewer: Does "decrypt" here mean just deriving the symmetric key and decrypting the ciphertext, or the full `DecryptNote` procedure of  [PROTOCOL] §4.10.2, including checking the note commitment?
How does this affect `valid` of the resulting JSON?

Reviewer: If the note commitment is note verified, attacks are possible (see discussing starting https://github.com/zcash/zcash/issues/2036#issuecomment-329864311).

Reviewer: ... and PR #2159 does not verify the note commitment, hence is vulnerable.

Author: Updated with a step 5 to verify note commitment.  Good point about "decrypt" and "DecryptNote", will think about this a bit more and perhaps update in a later draft.

Reviewer: In general, when specifying the actual algorithms, it would be good to use the same notation and level of formality as in the protocol spec. (For one, this would have caught this bug...)

::

  ZIP: 303
  Title: Payment Disclosure
  Owners: Deirdre Connolly <deirdre@zfnd.org>
  Original-Author: Simon Liu
  Status: Implemented (zcashd)
  Category: Informational
  Created: 2017-02-22
  License: MIT

Abstract
--------

This ZIP describes a method of proving that a payment was sent to a shielded address.  In the typical case, this means enabling a sender to present a proof that they transferred funds to a recipient's shielded address.  The method described will be compatible with the Zcash Protocol [#protocol]_ deployed at the launch of the Zcash network. See zcash issues `2036 <https://github.com/zcash/zcash/issues/2036>`_ and `737 <https://github.com/zcash/zcash/issues/737>`_ for context.

> Motivation

Reviewer 5: Another motivation is to simply share visibility of a shielded transaction with a third party.

----------
Payment disclosure is useful in a number of situations, for example:

* A sender may need to prove that their payment was sent and received by a recipient.  For example, a customer paid too much for an item and would like to claim a refund from the vendor.

> * A third party needs to verify that a payment between sender and recipient was executed successfully.  For example, a regulator needs to confirm a transfer of funds between two parties.

Reviewer 5: Note that the _recipient_ of a shielded transaction may wish to share visibility of a shielded transaction with a third party. The ZIP should specify whether a payment disclosure can be created by the recipient and, if so, how.

Reviewer 6: I think the recipient could only create a disclosure in the `transparent --> shielded` case. In other cases, they don't have the information required for the proof. However, I wonder about the motivation for a recipient-disclosed payment.

Bob retrieves his payment disclosure and emails it to Alice.  Alice is travelling and only has access to her smartphone.  She visits a Zcash blockchain explorer and copies the payment disclosure into the search box. The explorer confirms that Bob paid his share.

Charlie attempts to retrieve the payment disclosure but can't find one.  Charlie had used the command line to send his funds to Alice, but never checked the status of the z_sendmany operation.  Checking the debug logs he discovers that the operation failed due to an incorrect parameter.  Charlie tells Alice that he will re-send his payment.

Design Considerations
---------------------
The payment disclosure does not prove that the party presenting the payment disclosure is the sender.

To prevent a man-in-the-middle attack, the recipient could pose an interactive challenge involving some out-of-band secret which only the sender would pass successfully.

> It's also possible, rather than an interactive challenge, to make the disclosure dependent on data such as a refund address. This would prevent an attack where the payment disclosure is replayed with a claim that the refund should be to another address.

Reviewer 5: Idea: Add an optional field to a payment disclosure for a comment signed by the JoinSplitSig private key?

        }

#. Return to the caller the hex string of the serialized PaymentDisclosure.  If there were errors generating the payment disclosure, return a standard JSON-RPC error with an appropriate error message.

::

  00171deabcd9a66c9810ea926c0828e24f2ce880c0796ef5e5803252e8e2a2d609d16971a56a817d7b272d65395aab6910defc70e401a6e264aa642f2fe59f5168000000000000000000f4fda3737075b8b7f00715a5fcf106c74b75150d8f5578b973417f529d3464e2df1d76d937887b8aad659a4a0bdedb34a2706759bd64fa923863094eba504d7b05686f776479c6855c4eee0e2601301bea503ad96d4216702baa8db2141530ae58875def42590afb9681c22948dd2affba1bddd81eeafef1579a760bf13e8afd849287d30800

This raw hex string can be given to the recipient or a third party to use with the new RPC call ``z_validatepaymentdisclosure``:

``z_validatepaymentdisclosure paymentdisclosure``

To validate a payment disclosure, perform the following steps:

1. Deserialize the raw hex string into a ``PaymentDisclosure`` object
2. Retrieve the ``joinSplitPubKey`` for the transaction and verify the payment disclosure signature ``payloadSig``.
3. Retrieve the note ciphertext from the blockchain for ``txid``, ``js``, ``n``.

> 4. Use the ``esk`` to decrypt the ciphertext into plain text

Reviewer 2: ```suggestion
4. Use the ``esk`` to decrypt the ciphertext into plain text, where `decrypt` is the same operation defined in [#protocol]_ §4.16.2 Decryption (Sprout).
```

::

  ZIP: 303
  Title: Payment Disclosure
  Owners: Deirdre Connolly <deirdre@zfnd.org>
  Original-Author: Simon Liu
  Status: Implemented (zcashd)
  Category: Informational
  Created: 2017-02-22
  License: MIT

Abstract
--------

This ZIP describes a method of proving that a payment was sent to a shielded address.  In the typical case, this means enabling a sender to present a proof that they transferred funds to a recipient's shielded address.  The method described will be compatible with the Zcash Protocol [#protocol]_ deployed at the launch of the Zcash network. See zcash issues `2036 <https://github.com/zcash/zcash/issues/2036>`_ and `737 <https://github.com/zcash/zcash/issues/737>`_ for context.

> 
> Copyright Notice
> ----------------
> This ZIP is published under the MIT License.

Reviewer 2: ```suggestion
```

This document follows the common conventions defined for ZIPs in [ZIP-1].

.. [PROTOCOL] Zcash Protocol Specification: https://github.com/zcash/zips/blob/master/protocol/protocol.pdf
.. [ZIP-1] https://github.com/zcash/zips/blob/master/zips/0001.rst

Copyright Notice
----------------
This ZIP is published under the MIT License.

Motivation
----------
Payment disclosure is useful in a number of situations, for example:

* A sender may need to prove that their payment was sent and received by a recipient.  For example, a customer paid too much for an item and would like to claim a refund from the vendor.
* A third party needs to verify that a payment between sender and recipient was executed successfully.  For example, a regulator needs to confirm a transfer of funds between two parties.

When a transaction involves only transparent addresses, the sender and recipient can verify payment by examining the blockchain.  A third party can also perform this verification if they know the transparent addresses of the involved parties.

> However, if the transaction involves shielded addresses, the blockchain by itself does not contain enough information from which a record of the payment can be reconstructed and verified.

Reviewer 4: Grammar nit: "does not contain enough information to allow a record of the payment to be reconstructed and verified."

Author: ok

Let us examine the types of transaction that might occur and when a method of payment disclosure would be useful:

transparent --> transparent
  The source, destination and amount are visible on the public blockchain.  Payment disclosure is not required.

transparent --> shielded
  The destination address and amount sent to that address cannot be confirmed. Payment disclosure is required.

shielded --> transparent
  The recipient can see the amount received at their destination address, but cannot identify the sender.  Payment disclosure is required.

shielded --> shielded
  The sender, recipient and amount are unknown.  Payment disclosure required.

> TODO: Perhap use a table to make this easier to understand, from perspective of sender,recipient and third party?

Reviewer 4: I think this is fine as-is.

Author: ok

The next day Bob sends 1 ZEC to Alice's shielded address for his share of the bill.  Other coworkers do the same.  A week later Alice remembers that she is owed money so she checks her shielded address.  Alice sees that the balance is less than what it should be, so she asks her coworkers to confirm their payments.

Bob retrieves his payment disclosure and emails it to Alice.  Alice is travelling and only has access to her smartphone.  She visits a Zcash blockchain explorer and copies the payment disclosure into the search box. The explorer confirms that Bob paid his share.

Charlie attempts to retrieve the payment disclosure but can't find one.  Charlie had used the command line to send his funds to Alice, but never checked the status of the z_sendmany operation.  Checking the debug logs he discovers that the operation failed due to an incorrect parameter.  Charlie tells Alice that he will re-send his payment.

Design Considerations
---------------------
The payment disclosure does not prove that the party presenting the payment disclosure is the sender.

> To mitigate against a replay attack, the recipient could enforce a policy where the first person to present a given payment disclosure is the honest sender.

Reviewer 4: Delete this sentence; that's not a reliable way of doing it. It could be okay if it's sent over a secure channel but it's tricky to explain the resulting security properties.

Author: ok

A non-interactive solution is also possible:

#. The sender includes data, such as a refund address, in the memo field of a payment to a shielded address.  The memo field is non-malleable as it is part of the data signed by the sender when creating JoinSplitSig [PROTOCOL] §4.1.6 Signatures.  Currently, the public key to verify the signature is stored in the blockchain along with the transaction.
#. When presenting the payment disclosure to a recepient, the sender signs the payment disclosure with the JoinSplitSig private key which only they would know.  The JoinSplitSig key pair is ephemeral and normally discarded but the method described in this ZIP will retain the private portion.  The recipient can then use the public key to verify the signed disclosure.

Using the JoinSplitSig key pair in this way also means that a sender can make payments without having to first provide a refund address or other details at the time of the original payment.  If at a later date the sender and recipient need to confirm the payment, the sender can sign and send any data required to the recipient, who can in turn verify it.

Known Issues
------------

> Both plaintext outputs of a joinsplit are currently encrypted with symmetric keys derived from the same ephemeral secret key, as discusshed here: https://github.com/zcash/zcash/issues/558#issuecomment-167819936

Reviewer 4: s/discusshed/discussed/

Author: Ok, removing Sean Connery...

> This means that a payment disclosure which includes the ephemeral secret key intended to decipher the note ciphertext belonging to a particular joinsplit output, creates a privacy leak by also deciphering the other joinsplit output and revealing it's amount and payment address.

Reviewer 4: s/it's/its/

Reviewer 4: s/which/that/

Reviewer 4: its contents (amount, payment address, and memo field).

Author: ok

Note the current implementation of the ``z_sendmany`` RPC call:

- When sending from a transparent address to shielded addresses, each shielded address will occupy one output of a joinsplit.  Thus a payment address to a different recipient could be revealed.

- When sending involves chained joinsplits, for each joinsplit, one output is for payment to a recipient's address, while the other output is used as change back to the sender.  Thus the change address will be revealed, which is the sender's own shielded address.

A proposal to prevent information leakage of change addresses is currently under development [KDFT].  TODO: Does this block ZIP approval?

> Also note that an independent third party cannot know for sure if the sender and recipient are not colluding to hide value transfer.  JoinSplits have two inputs and two outputs.  The recipient may identify their receiving address for one output, but not disclose the fact that the other output also belongs to them.

Reviewer 4: s/for sure if/for sure that/

Reviewer 4: This seems to be moot because the sender and recipient could just as well transfer value in a different joinsplit without disclosing it.

Author: Updated note with this.

When a shielded transaction is created sucessfully and accepted into the local mempool with ``z_sendmany`` or ``z_shieldcoinbase``, the payment disclosure database is updated with information necessary to create a payment disclosure.  However, the transaction itself may never be mined and confirmed in the blockchain, rendering the database entry itself redundant and available for purging at a later date e.g. garbage collection. 

.. [KDFT] https://github.com/zcash/zcash/issues/2102

Specification
-------------
When creating a shielded transaction, for each JoinSplit output, a data structure is created to record the following fields:

- transaction id
- index [0..len-1] of JoinSplit in array of JoinSplits contained in transaction
- index [0..1] of JoinSplit output
- recipient's payment address is a shielded address ``(a_pk, pk_enc)`` [PROTOCOL] §3.1 Payment Addresses and Keys

> - symmetric key used to encrypt the note, also referred to as the ephemeral secret key ``esk`` [PROTOCOL] §4.10.1 Generate a new KA (public, private) key pair ``(epk, esk)``.

Reviewer 4: ``esk`` is not a symmetric key, it's a private key. In general avoid saying "ephemeral secret key" because that's confusing. Instead say:

* ephemeral private key ``esk`` used to encrypt the note

Author: ack

- JoinSplitSig private key used to sign the JoinSplit transaction ``joinSplitPrivKey`` [PROTOCOL] §4.4 Sending Notes

The payment disclosure data should be persisted to disk or a database so it can be retrieved later.

When persisting, third party applications should expect serialization of payment disclosure to follow Zcash and upstream convention.

::

    ADD_SERIALIZE_METHODS;

>     ...

Reviewer 4: Fill in the dots.

Author: ok

    READWRITE(txid);
    ...
    READWRITE(joinSplitPrivKey);

A new RPC call will be introduced to allow the sender to retrieve the payment disclosure data for a given shielded output index.

``z_getpaymentdisclosure txid joinsplit_index output_index [message]``

> - Returns the payment disclosure in hexadecimal format.

Reviewer 4: Say that hexadecimal digits are case-insensitive.

Author: ok

      "memo": "f600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "value": 10.00000000,
      "commitmentMatch": true,
      "valid": true
    }

Valid field is true if all the following conditions hold:

> 

Reviewer 4: - the payment disclosure is correctly encoded and has a recognized version

Author: ok

- txid is confirmed in the blockchain
- jsindex is valid for the txid
- outputindex is within range [0..1]
- paymentaddress is a valid shielded address for the network we are on
- value is within range [0..MAX_MONEY]

> - message is within constraints e.g. number of characters, allowed characters

Reviewer 4: Do we actually check allowed characters? I think we do not and should not.

Author: Currently there are no constraints.  We should consider what our max message length is and if there is any restriction on the characters.

::

  ZIP: 303
  Title: Payment Disclosure
  Owners: Deirdre Connolly <deirdre@zfnd.org>
  Original-Author: Simon Liu
  Status: Implemented (zcashd)
  Category: Informational
  Created: 2017-02-22
  License: MIT

> ==================
> Payment Disclosure
> ==================

Reviewer 2: ```suggestion
```

      "value": 10.00000000,
      "commitmentMatch": true,
      "valid": true
    }

Valid field is true if all the following conditions hold:

- the payment disclosure is correctly encoded and has a recognized version
- txid is confirmed in the blockchain
- jsindex is valid for the txid
- outputindex is within range [0..1]
- paymentaddress is a valid shielded address for the network we are on
- value is within range [0..MAX_MONEY]

> - message is within constraints (TODO: currently there are no constrainst defined, such as number of characters, allowed characters)

Reviewer 2: ```suggestion
- message is within constraints (TODO: currently there are no constraints defined, such as number of characters, allowed characters)
```

- signature is valid for all of the above fields
- commitment derived from the deciphered note matches the commitment in the blockchain

Otherwise an error field is returned explaining why the payment disclosure is invalid:

::

    {
    valid : false,
    ...

>     error : "The payment disclosuse is invalid because..."

Reviewer 2: ```suggestion
    error : "The payment disclosure is invalid because..."
```

::

  ZIP: 303
  Title: Payment Disclosure
  Owners: Deirdre Connolly <deirdre@zfnd.org>
  Original-Author: Simon Liu
  Status: Implemented (zcashd)
  Category: Informational
  Created: 2017-02-22
  License: MIT

==================
Payment Disclosure
==================

Abstract
--------

> This ZIP describes a method of proving that a payment was sent to a shielded address.  In the typical case, this means enabling a sender to present a proof that they transferred funds to a recipient's shielded address.  The method described will be compatible with the Zcash Protocol [PROTOCOL] deployed at the launch of the Zcash network.

Reviewer 2: ```suggestion
This ZIP describes a method of proving that a payment was sent to a shielded address.  In the typical case, this means enabling a sender to present a proof that they transferred funds to a recipient's shielded address.  The method described will be compatible with the Zcash Protocol [#protocol]_ deployed at the launch of the Zcash network. See `zcash issue 2036 <https://github.com/zcash/zcash/issues/2036>`_ for context.
```

> This document follows the common conventions defined for ZIPs in [ZIP-1].
> 
> .. [PROTOCOL] Zcash Protocol Specification: https://github.com/zcash/zips/blob/master/protocol/protocol.pdf
> .. [ZIP-1] https://github.com/zcash/zips/blob/master/zips/0001.rst

Reviewer 2: ```suggestion
```

Note the current implementation of the ``z_sendmany`` RPC call:

- When sending from a transparent address to shielded addresses, each shielded address will occupy one output of a joinsplit.  Thus a payment address to a different recipient could be revealed.

- When sending involves chained joinsplits, for each joinsplit, one output is for payment to a recipient's address, while the other output is used as change back to the sender.  Thus the change address will be revealed, which is the sender's own shielded address.

A proposal to prevent information leakage of change addresses is currently under development [KDFT].  TODO: Does this block ZIP approval?

Note that an independent third party cannot know for sure that the sender and recipient are not colluding to hide value transfer.  JoinSplits have two inputs and two outputs.  The recipient may identify their receiving address for one output, but not disclose the fact that the other output also belongs to them.  Also, the sender and recipient could collude to transfer value in a different JoinSplit without disclosing it.

> When a shielded transaction is created sucessfully and accepted into the local mempool with ``z_sendmany`` or ``z_shieldcoinbase``, the payment disclosure database is updated with information necessary to create a payment disclosure.  However, the transaction itself may never be mined and confirmed in the blockchain, rendering the database entry itself redundant and available for purging at a later date e.g. garbage collection. 

Reviewer 2: ```suggestion
When a shielded transaction is created successfully and accepted into the local mempool with ``z_sendmany`` or ``z_shieldcoinbase``, the payment disclosure database is updated with information necessary to create a payment disclosure.  However, the transaction itself may never be mined and confirmed in the blockchain, rendering the database entry itself redundant and available for purging at a later date e.g. garbage collection. 
```

.. [KDFT] https://github.com/zcash/zcash/issues/2102

Specification
-------------
When creating a shielded transaction, for each JoinSplit output, a data structure is created to record the following fields:

- transaction id
- index [0..len-1] of JoinSplit in array of JoinSplits contained in transaction
- index [0..1] of JoinSplit output

> - recipient's payment address is a shielded address ``(a_pk, pk_enc)`` [PROTOCOL] §3.1 Payment Addresses and Keys
> - ephemeral private key ``esk`` used to encrypt the note [PROTOCOL] §4.10.1 Generate a new KA (public, private) key pair ``(epk, esk)``.
> - JoinSplitSig private key used to sign the JoinSplit transaction ``joinSplitPrivKey`` [PROTOCOL] §4.4 Sending Notes

Reviewer 2: ```suggestion
- recipient's payment address is a shielded address ``(a_pk, pk_enc)`` [#protocol]_ §3.1 Payment Addresses and Keys
- ephemeral private key ``esk`` used to encrypt the note [#protocol]_ §4.10.1 Generate a new KA (public, private) key pair ``(epk, esk)``.
- JoinSplitSig private key used to sign the JoinSplit transaction ``joinSplitPrivKey`` [#protocol]_ §4.4 Sending Notes
```

3. Retrieve the note ciphertext from the blockchain for ``txid``, ``js``, ``n``.
4. Use the ``esk`` to decrypt the ciphertext into plain text
5. Derive commitment from plain text and check it matches commitment in blockchain
6. Return JSON output as described above in the specification. 

Possible error messages which could cause validation to fail include:

TODO: Add bad prefix error message here
- "Invalid parameter, expected payment disclosure data in hexadecimal format."
- "Invalid parameter, payment disclosure data is malformed."
- "No information available about transaction"
- "Transaction has not been confirmed yet"
- "Transaction is not a shielded transaction"
- "Payment disclosure refers to an invalid joinsplit index"
- "Payment disclosure refers to an invalid output index"
- "Payment disclosure refers to an unknown version"
- "Payment disclosure signature does not match transaction signature"
- "Payment disclosure refers to an invalid payment address"
- "Payment disclosure derived commitment does not match blockchain commitment"
- "Payment disclosure error when deciphering note"
- ...

> References
> ----------
> | https://github.com/zcash/zcash/issues/2036
> | https://github.com/Electric-Coin-Company/zecc-private/wiki/Engineering-meeting-notes
> | https://github.com/zcash/zcash/issues/737

Reviewer 2: ```suggestion
References
==========

.. [#protocol] `Zcash Protocol Specification, Version 2020.1.5 or later <protocol/protocol.pdf>`_
```

| Title: Payment Disclosure

> | ZIP: 00? (DRAFT)

Reviewer 2: ```suggestion
| ZIP: 303
```

> | Title: Payment Disclosure
> | ZIP: 00? (DRAFT)
> | Replaces: n/a
> | Superseded-by: n/a
> | Editor: Simon Liu <simon@z.cash>

Reviewer 2: ```suggestion
::

  ZIP: 303
  Title: Payment Disclosure
  Owner: Deirdre Connolly <deirdre@zfnd.org>
  Original-Author: Simon Liu
  Status: Implemented (zcashd)
  Category: Informational
  Created: 2017-02-22
  License: MIT
```

Charlie attempts to retrieve the payment disclosure but can't find one.  Charlie had used the command line to send his funds to Alice, but never checked the status of the z_sendmany operation.  Checking the debug logs he discovers that the operation failed due to an incorrect parameter.  Charlie tells Alice that he will re-send his payment.

Design Considerations
---------------------
The payment disclosure does not prove that the party presenting the payment disclosure is the sender.

[TODO: Example?] To prevent a man-in-the-middle attack, the recipient could pose an interactive challenge involving some out-of-band secret which only the sender would pass successfully.

A non-interactive solution is also possible:

#. The sender includes data, such as a refund address, in the memo field of a payment to a shielded address.  The memo field is non-malleable as it is part of the data signed by the sender when creating JoinSplitSig [PROTOCOL] §4.1.6 Signatures.  Currently, the public key to verify the signature is stored in the blockchain along with the transaction.

> #. When presenting the payment disclosure to a recepient, the sender signs the payment disclosure with the JoinSplitSig private key which only they would know.  The JoinSplitSig key pair is ephemeral and normally discarded but the method described in this ZIP will retain the private portion.  The recipient can then use the public key to verify the signed disclosure.

Reviewer 7: Typo: "recepient"

> #. When presenting the payment disclosure to a recepient, the sender signs the payment disclosure with the JoinSplitSig private key which only they would know.  The JoinSplitSig key pair is ephemeral and normally discarded but the method described in this ZIP will retain the private portion.  The recipient can then use the public key to verify the signed disclosure.

Reviewer 7: ```suggestion
#. When presenting the payment disclosure to a recipient, the sender signs the payment disclosure with the JoinSplitSig private key which only they would know.  The JoinSplitSig key pair is ephemeral and normally discarded but the method described in this ZIP will retain the private portion.  The recipient can then use the public key to verify the signed disclosure.
```

Note the current implementation of the ``z_sendmany`` RPC call:

- When sending from a transparent address to shielded addresses, each shielded address will occupy one output of a joinsplit.  Thus a payment address to a different recipient could be revealed.

- When sending involves chained joinsplits, for each joinsplit, one output is for payment to a recipient's address, while the other output is used as change back to the sender.  Thus the change address will be revealed, which is the sender's own shielded address.

A proposal to prevent information leakage of change addresses is currently under development [KDFT].  TODO: Does this block ZIP approval?

Note that an independent third party cannot know for sure that the sender and recipient are not colluding to hide value transfer.  JoinSplits have two inputs and two outputs.  The recipient may identify their receiving address for one output, but not disclose the fact that the other output also belongs to them.  Also, the sender and recipient could collude to transfer value in a different JoinSplit without disclosing it.

> When a shielded transaction is created sucessfully and accepted into the local mempool with ``z_sendmany`` or ``z_shieldcoinbase``, the payment disclosure database is updated with information necessary to create a payment disclosure.  However, the transaction itself may never be mined and confirmed in the blockchain, rendering the database entry itself redundant and available for purging at a later date e.g. garbage collection. 

Reviewer 7: Typo: "sucessfully"

> When a shielded transaction is created sucessfully and accepted into the local mempool with ``z_sendmany`` or ``z_shieldcoinbase``, the payment disclosure database is updated with information necessary to create a payment disclosure.  However, the transaction itself may never be mined and confirmed in the blockchain, rendering the database entry itself redundant and available for purging at a later date e.g. garbage collection. 

Reviewer 7: ```suggestion
When a shielded transaction is created successfully and accepted into the local mempool with ``z_sendmany`` or ``z_shieldcoinbase``, the payment disclosure database is updated with information necessary to create a payment disclosure.  However, the transaction itself may never be mined and confirmed in the blockchain, rendering the database entry itself redundant and available for purging at a later date e.g. garbage collection. 
```

      "value": 10.00000000,
      "commitmentMatch": true,
      "valid": true
    }

Valid field is true if all the following conditions hold:

- the payment disclosure is correctly encoded and has a recognized version
- txid is confirmed in the blockchain
- jsindex is valid for the txid
- outputindex is within range [0..1]
- paymentaddress is a valid shielded address for the network we are on
- value is within range [0..MAX_MONEY]

> - message is within constraints (TODO: currently there are no constrainst defined, such as number of characters, allowed characters)

Reviewer 7: Typo: "constrainst"

> - message is within constraints (TODO: currently there are no constrainst defined, such as number of characters, allowed characters)

Reviewer 7: ```suggestion
- message is within constraints (TODO: currently there are no constraints defined, such as number of characters, allowed characters)
```

- signature is valid for all of the above fields
- commitment derived from the deciphered note matches the commitment in the blockchain

Otherwise an error field is returned explaining why the payment disclosure is invalid:

::

    {
    valid : false,
    ...

>     error : "The payment disclosuse is invalid because..."

Reviewer 7: Typo: "disclosuse"

>     error : "The payment disclosuse is invalid because..."

Reviewer 7: ```suggestion
    error : "The payment disclosure is invalid because..."
```

If the sender needs to provide a payment disclosure to the recipient or a third party, the sender will use RPC call ``z_getpaymentdisclosure`` to generate a Payment Disclosure.

``z_getpaymentdisclosure txid joinsplit_index output_index [message]``

To create a valid Payment Disclosure an implementation must:

#. Check the txid was confirmed in the blockchain
#. Create a ``PaymentDisclosureKey`` from parameters to RPC call
#. Use the key to retrieve its value ``PaymentDisclosureInfo`` from storage
#. Create and populate a ``PaymentDisclosure``

    ::
    
        struct PaymentDisclosurePayload {
            int32_t marker = PAYMENT_DISCLOSURE_PAYLOAD_MAGIC_BYTES;  // to be disjoint from transaction encoding
            uint8   version;        // 0 = experimental, 1 = first production version, etc.
            uint256 esk;            // zcash/NoteEncryption.cpp
            uint256 txid;           // primitives/transaction.h
            size_t js;              // Index into CTransaction.vjoinsplit
            uint8_t n;              // Index into JSDescription fields of length ZC_NUM_JS_OUTPUTS

>             PaymentAddress zaddr;   // zcash/Address.hpp

Reviewer 5: @daira Isn't `PaymentAddress` the recipient's address?

The next day Bob sends 1 ZEC to Alice's shielded address for his share of the bill.  Other coworkers do the same.  A week later Alice remembers that she is owed money so she checks her shielded address.  Alice sees that the balance is less than what it should be, so she asks her coworkers to confirm their payments.

Bob retrieves his payment disclosure and emails it to Alice.  Alice is travelling and only has access to her smartphone.  She visits a Zcash blockchain explorer and copies the payment disclosure into the search box. The explorer confirms that Bob paid his share.

Charlie attempts to retrieve the payment disclosure but can't find one.  Charlie had used the command line to send his funds to Alice, but never checked the status of the z_sendmany operation.  Checking the debug logs he discovers that the operation failed due to an incorrect parameter.  Charlie tells Alice that he will re-send his payment.

Design Considerations
---------------------
The payment disclosure does not prove that the party presenting the payment disclosure is the sender.

> [TODO: Example?] To prevent a man-in-the-middle attack, the recipient could pose an interactive challenge involving some out-of-band secret which only the sender would pass successfully.

Reviewer 2: ```suggestion
To prevent a man-in-the-middle attack, the recipient could pose an interactive challenge involving some out-of-band secret which only the sender would pass successfully.
It's also possible, rather than an interactive challenge, to make the disclosure dependent on data such as a refund address. This would prevent an attack where the payment disclosure is replayed with a claim that the refund should be to another address.
```

Bob retrieves his payment disclosure and emails it to Alice.  Alice is travelling and only has access to her smartphone.  She visits a Zcash blockchain explorer and copies the payment disclosure into the search box. The explorer confirms that Bob paid his share.

Charlie attempts to retrieve the payment disclosure but can't find one.  Charlie had used the command line to send his funds to Alice, but never checked the status of the z_sendmany operation.  Checking the debug logs he discovers that the operation failed due to an incorrect parameter.  Charlie tells Alice that he will re-send his payment.

Design Considerations
---------------------
The payment disclosure does not prove that the party presenting the payment disclosure is the sender.

To prevent a man-in-the-middle attack, the recipient could pose an interactive challenge involving some out-of-band secret which only the sender would pass successfully.
It's also possible, rather than an interactive challenge, to make the disclosure dependent on data such as a refund address. This would prevent an attack where the payment disclosure is replayed with a claim that the refund should be to another address.

> A non-interactive solution is also possible:
> 
> #. The sender includes data, such as a refund address, in the memo field of a payment to a shielded address.  The memo field is non-malleable as it is part of the data signed by the sender when creating JoinSplitSig [PROTOCOL] §4.1.6 Signatures.  Currently, the public key to verify the signature is stored in the blockchain along with the transaction.
> #. When presenting the payment disclosure to a recepient, the sender signs the payment disclosure with the JoinSplitSig private key which only they would know.  The JoinSplitSig key pair is ephemeral and normally discarded but the method described in this ZIP will retain the private portion.  The recipient can then use the public key to verify the signed disclosure.
> 
> Using the JoinSplitSig key pair in this way also means that a sender can make payments without having to first provide a refund address or other details at the time of the original payment.  If at a later date the sender and recipient need to confirm the payment, the sender can sign and send any data required to the recipient, who can in turn verify it.
> 

Reviewer 2: ```suggestion

```

However, if the transaction involves shielded addresses, the blockchain by itself does not contain enough information to allow a record of the payment to be reconstructed and verified.

Let us examine the types of transaction that might occur and when a method of payment disclosure would be useful:

transparent --> transparent
  The source, destination and amount are visible on the public blockchain.  Payment disclosure is not required.

transparent --> shielded
  The destination address and amount sent to that address cannot be confirmed. Payment disclosure is required.

shielded --> transparent
  The recipient can see the amount received at their destination address, but cannot identify the sender.  Payment disclosure is required.

shielded --> shielded
  The sender, recipient and amount are unknown.  Payment disclosure required.

Requirements
------------

Allow the sender to automatically or optionally create a payment disclosure when creating a transaction involving shielded addresses.

The sender or a third party can present this payment disclosure to the recipient to prove that funds were sent to their address.

> The mechanism should support both payments involving either the Sprout or Sapling protocols.

Reviewer 6: This could remove Sprout and should add Orchard.

Known Issues
------------

Both plaintext outputs of a Sprout JoinSplit are currently encrypted with symmetric keys derived from the same ephemeral secret key, as discussed here: https://github.com/zcash/zcash/issues/558#issuecomment-167819936

This means that a payment disclosure that includes the ephemeral secret key intended to decipher the note ciphertext belonging to a particular JoinSplit output, creates a privacy leak by also deciphering the other JoinSplit output and revealing its contents (amount, payment address and memo field).

Note the current implementation of the ``z_sendmany`` RPC call:

- When sending from a transparent address to shielded addresses, each shielded address will occupy one output of a JoinSplit.  Thus a payment address to a different recipient could be revealed.

- When sending involves chained JoinSplits, for each JoinSplit, one output is for payment to a recipient's address, while the other output is used as change back to the sender.  Thus the change address will be revealed, which is the sender's own shielded address.

> A proposal to prevent information leakage of change addresses is currently under development [KDFT].  TODO: Does this block ZIP approval?

Reviewer 6: All of the Known Issues to this point are Sprout-specific, and while there is already experimental support for it in zcashd, Sprout doesn't need to have better (or any support) for payment disclosures. It could be restricted to Sapling and Orchard which don't have these issues.

    and auxiliary input :math:`(path, 0, g_d, pk_d, 1, 0, cm, 0, α, ak, nsk)`.
    [#spend-statement]_

  - Let :math:`rsk = \mathsf{SpendAuthSig.RandomizePrivate}(α, ask)`.

  - Let :math:`coinType` be the 4-byte little-endian encoding of the coin type in its index
    form, not its hardened form (i.e. 133 for mainnet Zcash).

  - Let :math:`digest = \mathsf{BLAKE2b}\text{-}\mathsf{256}(\texttt{"ZIP304Signed"}\,||\,coinType, zkproof\,||\,msg)`.

  - Let :math:`spendAuthSig = \mathsf{SpendAuthSig.Sign}(rsk, digest)`.

- Return :math:`(d, pk_d, nf, zkproof, spendAuthSig)`.

The payment disclosure data should be persisted to disk or a database so it can be retrieved later.

When persisting, third party applications should expect serialization of payment disclosure to follow Zcash and upstream convention.

::

    ADD_SERIALIZE_METHODS;
    ...
    
    READWRITE(marker);
    READWRITE(version);
    READWRITE(esk);
    READWRITE(txid);
    READWRITE(js);
    READWRITE(n);
    READWRITE(zaddr);
    READWRITE(message);

A new RPC call will be introduced to allow the sender to retrieve the payment disclosure data for a given shielded output index.

> ``z_getpaymentdisclosure txid joinsplit_index output_index [message]``

Reviewer 6: The `joinsplit_index output_index` here isn't sufficiently expressive. What about transparent, Sapling, and Orchard outputs?

Also, the sender may not need to reveal all inputs to show that sufficient value has been transferred. Maybe include an optional list of input indices to reveal in the disclosure.

Do Orchard actions have a similar issue to JoinSplits, where revealing part of it could also reveal the input, even if it's not one the sender wants to reveal?
